### PR TITLE
feature(shoppingCart): adds shopping cart context

### PR DESCRIPTION
**Pull request by RZhyvitskyi**
# Description:

Context works in all places. 

I have only one issue with deleting `order items` from cart if some of the orders is already selected.

# Jira:

https://hackyourfuture-dk.atlassian.net/jira/software/projects/CLASS22/boards/5?selectedIssue=CLASS22-132

### CODE REVIEW

File: packages/client/src/containers/DropdownGroup/DropdownGroup.Container.js [hunk 1]
```diff
@@ -14,11 +14,12 @@ export const DropdownGroup = ({ productId }) => {
     const fetchVariantsByProductId = async () => {
       const response = await fetch(`${apiURL()}/variants?product=${productId}`);
       const variantsProductData = await response.json();
+
       setVariants(variantsProductData);
     };
 
     fetchVariantsByProductId();
-  }, [productId, variants]);
+  }, [productId]);
```
Comment (miltonhultgren) [inline packages/client/src/containers/DropdownGroup/DropdownGroup.Container.js:24]: Why was this changed?

File: packages/client/src/components/ViewPageButton/ViewPageButton.component.js [hunk 2]
```diff
@@ -2,9 +2,12 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import './ViewPageButton.styles.css';
 
-export const ViewPageButton = ({ label, backgroundColor }) => {
+export const ViewPageButton = ({ label, backgroundColor, onClick }) => {
```
Comment (miltonhultgren) [inline packages/client/src/components/ViewPageButton/ViewPageButton.component.js]: I'm not sure I understand why this is needed?

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  useCallback,
+  createContext,
+} from 'react';
+import PropTypes from 'prop-types';
+import { useUserContext } from '../../userContext';
+
+const ShoppingCartContext = createContext({});
+
+export const useShoppingCart = () => {
+  return useContext(ShoppingCartContext);
```
Comment (miltonhultgren) [inline packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx:16]: Here you should check if the context actually has a value because it might not have been wrapped properly.

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      setPendingItems(updatedPendingItems);
+    },
+    [pendingItems, user?.uid],
+  );
+
+  const shoppingCartProviderValue = useMemo(
+    () => ({
+      pendingItems,
+      deletePendingItem,
+      addPendingItem,
```
Comment (miltonhultgren) [inline packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: What do you mean by `pending`? The word confuses me :D You mean in the sense that they have yet to be ordered and paid for? 
I would remove that word because I think it adds more confusion than it adds clarity. The fact that they are in the cart already tells me they have not yet been ordered.

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+export const useShoppingCart = () => {
+  return useContext(ShoppingCartContext);
+};
+
+const ShoppingCartProvider = ({ children }) => {
+  const { user } = useUserContext();
+
+  const [pendingItems, setPendingItems] = useState(() => {
+    const items = localStorage.getItem(user?.uid);
```
Comment (miltonhultgren) [inline packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: Maybe you can sync up with @Yuliia-Balandiuk to talk about creating a custom hook for local storage.

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    (itemId) => {
+      const filteredItems = pendingItems.filter(({ id }) => id !== itemId);
+      localStorage.setItem(user?.uid, JSON.stringify(filteredItems));
+      setPendingItems(filteredItems);
+    },
+    [pendingItems, user?.uid],
+  );
+
+  const addPendingItem = useCallback(
+    (test) => {
```
Comment (miltonhultgren) [inline packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: Maybe a better name than test here? :)

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      const updatedPendingItems = { ...pendingItems, ...test };
```
Comment (miltonhultgren) [inline packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: `pendingItems` is a list, but here you spread it into an object, that's probably a mistake.
Further, `test` is probably and object representing a product so I don't think you want to spread that apart.

**Review by miltonhultgren — DISMISSED** [review]

File: packages/client/src/containers/DropdownGroup/DropdownGroup.Container.js (hunk 1 above)
Comment (RZhyvitskyi) [reply packages/client/src/containers/DropdownGroup/DropdownGroup.Container.js:24]: It was causing an infinite loop

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx (hunk 5 above)
Comment (RZhyvitskyi) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: We have changed mind about ShoppingCartContext, it will be handled with database

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx (hunk 6 above)
Comment (RZhyvitskyi) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: Hahaha, true 👍

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx (hunk 4 above)
Comment (RZhyvitskyi) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: Make sense 👍 I'll change it to orderItems soon

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx (hunk 5 above)
Comment (miltonhultgren) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: You'll put the users current shopping cart into the database as they are preparing it?
That doesn't make a lot of sense in my opinion. Imagine if they stop midway before making an order, you'll end up with a lot of junk in the database. 

For this kind of data, local storage is perfect. The actual order should of course end up in the database once it's confirmed and paid for.

Comment (RZhyvitskyi) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: If we look on web shops which provides best service, they keep cart data in database, Amazon one of them. Space for data is very cheap in our days. So I'm vouting for database cart data.

Comment (RZhyvitskyi) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: But you are right, we do need to use localstorage for unauthorized users

**Comment by shpomp** [comment]
👀

File: packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx (hunk 5 above)
Comment (miltonhultgren) [reply packages/client/src/utils/ShoppingCartContext/ShoppingCartContext.jsx]: Space is cheap, but it's mostly cheap **at scale**. There is a million and one things Amazon does that do not apply to this application. 

While I can agree with your point of "best service", there is a lot more that goes into that. Amazon has much more complex systems in place to manage that data, such as expiry dates on incomplete orders, not to mention the amount of effort they have put into making their sales funnel streamlined so that a large majority of shopping carts actually get turned into orders. 

It's good to be inspired, but it is critical to consider how your situation is different from that of the big tech companies, at least until you actually work for one of the big tech companies. Amazon doesn't use MySQL to store this kind of data and at the scale your solving for right now, you don't need such complexity, and in some situations you cannot even afford such complexity. Amazon did not start out huge, they grew!

**Review by viedamayanti — DISMISSED** [review]

**Review by ShravyaGurmitkal — DISMISSED** [review]

**Review by viedamayanti — APPROVED** [review]

File: packages/client/src/components/ViewPageButton/ViewPageButton.component.js (hunk 2 above)
Comment (RZhyvitskyi) [reply packages/client/src/components/ViewPageButton/ViewPageButton.component.js]: We can send a onClick handler function.
